effect/shared/src/main/scala/com/monovore/decline/effect/CommandIOApp.scala

```diff
@@ -14,34 +16,55 @@ abstract class CommandIOApp(
 
   def main: Opts[IO[ExitCode]]
 
-  private[this] def command: Command[IO[ExitCode]] = {
-    val showVersion = {
-      if (version.isEmpty) Opts.never
-      else {
-        val flag = Opts.flag(
-          long = "version",
-          short = "v",
-          help = "Print the version number and exit.",
-          visibility = Visibility.Partial
-        )
-        flag.as(IO(System.out.println(version)).as(ExitCode.Success))
-      }
-    }
+  override final def run(args: List[String]): IO[ExitCode] = {
+    import CommandIOApp._
+
+    val opts = if (version.isEmpty) main else addVersionFlag(main)(version)
+    val command = Command(name, header, helpFlag)(opts)
 
-    Command(name, header, helpFlag)(showVersion orElse main)
+    for {
+      parseResult <- IO(command.parse(PlatformApp.ambientArgs getOrElse args, sys.env))
+      exitCode <- parseResult.fold(printHelp[IO], identity)
+    } yield exitCode
   }
 
-  override final def run(args: List[String]): IO[ExitCode] = {
-    def printHelp(help: Help): IO[ExitCode] =
-      IO(System.err.println(help)).as {
-        if (help.errors.size > 0) ExitCode.Error
-        else ExitCode.Success
-      }
+}
+
+object CommandIOApp {
 
+  def run[F[_]](
```

Comment:
maybe expose `def command[F[_]]` and and just build the command up?

Reply:
It would require two function calls in client code but I guess that's ok

Reply:
I think taking this suggestion is less important if you remove the duplication from the class and the companion object. I was looking for opportunities for reuse.